### PR TITLE
Create model output endpoint

### PR DESCRIPTION
**Pull request by abhaasgoyal**
Related to #63 

Implement endpoint for creating a new model output (`POST` method) for a given Experiment ID + Model Profile ID

Add Github secrets for the following (following from the existing model output id)

1. `MEORG_MODEL_OUTPUT_NAME` - If no environment variable is provided for `pytest`, `"meorg-client-model-output"` is used
2. `MEORG_MODEL_PROFILE_ID`
3. `MEORG_EXPERIMENT_ID`

### CODE REVIEW

File: meorg_client/client.py [hunk 1]
```diff
@@ -442,6 +442,31 @@ def start_analysis(self, id: str) -> Union[dict, requests.Response]:
             url_params=dict(id=id),
         )
 
+    def post_output_entity(
```
Comment (bschroeter) [inline meorg_client/client.py:445]: The name `post_output_entity` doesn't follow the established naming convention.

Suggest something like:

```suggestion
    def model_output_create(
```

File: meorg_client/client.py [hunk 2]
```diff
@@ -442,6 +442,31 @@ def start_analysis(self, id: str) -> Union[dict, requests.Response]:
...
+    ) -> Union[dict, requests.Response]:
+        """
+        Create a new model output entity
+        Parameters
+        ----------
+        model_id : str
+            Model Profile ID
+        exp_id : str
+            Experiment ID
+        name : _type_
```
Comment (bschroeter) [inline meorg_client/client.py:456]:
```suggestion
        name : str
```

File: meorg_client/tests/test_client.py [hunk 3]
```diff
@@ -93,6 +93,14 @@ def test_list_endpoints(client: Client):
     assert isinstance(response, dict)
 
 
+def test_create_model_output(
+    client: Client, model_profile_id, experiment_id, model_output_name
+):
+    """Test Creation of Model output"""
+    _ = client.post_output_entity(model_profile_id, experiment_id, model_output_name)
```
Comment (bschroeter) [inline meorg_client/tests/test_client.py:100]: A further test of this would be to attempt to pull the details of the newly-created model output and ensure that it returns as expected.

Comment (bschroeter) [inline meorg_client/tests/test_client.py:100]: Update method name as above.

File: meorg_client/cli.py [hunk 4]
```diff
@@ -214,6 +214,42 @@ def analysis_start(id: str):
...
+        Experiment ID.
+
+    name : str
+        New model output name
+
+    Prints modeloutput ID of created object, and whether it already existed or not.
+    """
+    client = _get_client()
+
+    response = _call(
```
Comment (bschroeter) [inline meorg_client/cli.py:241]: Change method name as below.

File: meorg_client/cli.py [hunk 5]
```diff
@@ -214,6 +214,42 @@ def analysis_start(id: str):
...
+        client.post_output_entity, model_id=model_id, exp_id=exp_id, name=name
+    )
+
+    if client.success():
+        output_id = response.get("data").get("modeloutputId")
+        created = response.get("data").get("created")
```
Comment (bschroeter) [inline meorg_client/cli.py:247]: Can you pls explain how this works? Does it return a key in the data called "created" that is a new model output id?

File: meorg_client/cli.py [hunk 6]
```diff
@@ -291,6 +327,11 @@ def cli_analysis():
     pass
 
 
+@click.group("output", help="Model output commands.")
+def cli_output():
```
Comment (bschroeter) [inline meorg_client/cli.py:331]:
```suggestion
def cli_model_output():
```

File: meorg_client/cli.py [hunk 7]
```diff
@@ -304,6 +345,9 @@ def cli_analysis():
 cli_analysis.add_command(analysis_start)
 cli_analysis.add_command(analysis_status)
 
+# Add output command
+cli_output.add_command(create_new_model_output)
```
Comment (bschroeter) [inline meorg_client/cli.py:349]:
```suggestion
cli_model_output.add_command(create_new_model_output)
```

File: meorg_client/client.py (hunk 1 above)
Comment (bschroeter) [reply meorg_client/client.py:445]: The general idea (although I've failed to follow this myself on occasion) is:

`{THING}_{VERB}`, where VERB is anything other than a standard getting of the details.

So a model_output will eventually have:
- `model_output` - For getting details about a model output and usually operating over http `GET`
- `model_output_create` - For creating one and usually operating over http `POST`
- `model_output_delete` - For deleting one and usually operating over http `DELETE`.

There are the odd use-cases for http `PATCH` in the case of an update, but I don't think that will impact this part of the code.

File: meorg_client/endpoints.py [hunk 8]
```diff
@@ -16,3 +16,6 @@
 # Analysis
 ANALYSIS_START = "modeloutput/{id}/start"
 ANALYSIS_STATUS = "analysis/{id}/status"
+
+# Model Outputs
+MODEL_OUTPUT = "modeloutput"
```
Comment (bschroeter) [inline meorg_client/endpoints.py:21]: I expect this will change as the other methods are implemented? Or are they all implemented via the http methods on this one endpoint?

File: meorg_client/cli.py [hunk 9]
```diff
@@ -214,6 +214,42 @@ def analysis_start(id: str):
         click.echo(analysis_id)
 
 
+@click.command("create")
+@click.argument("model_id")
+@click.argument("exp_id")
+@click.argument("name")
+def create_new_model_output(model_id: str, exp_id: str, name: str):
+    """
+    Create a new model output entity
```
Comment (bschroeter) [inline meorg_client/cli.py:223]:
```suggestion
    Create a new model output profile.
```

**Review by bschroeter — CHANGES_REQUESTED** [review]
A few small things as indicated in the comments, most of them around the semantics of how things are currently named etc. This is mostly about future-proofing what we have and making it easier to enhance the client going forward.

Some other things that would be useful here:
- Update the docs to include details about how to use the new endpoint(s)
- More comprehensive tests, beyond just ensuring that a command doesn't return non-zero, but then using other parts of the client to ensure that what was created can indeed be returned as expected.

Nice work.

File: meorg_client/endpoints.py (hunk 8 above)
Comment (abhaasgoyal) [reply meorg_client/endpoints.py:21]: Considering that there are multiple endpoints related to model output operations.. For now, I think renaming it to `MODEL_OUTPUT_CREATE' should be good.

File: meorg_client/cli.py (hunk 5 above)
Comment (abhaasgoyal) [reply meorg_client/cli.py:247]: In this case, `created` is value indicating whether resource was existing or new (i.e. model output was already present for the input values

Comment (abhaasgoyal) [reply meorg_client/cli.py:247]: Although It is mentioned in the `echo` statements below what this means, should I still add a comment here?

File: meorg_client/tests/test_client.py (hunk 3 above)
Comment (abhaasgoyal) [reply meorg_client/tests/test_client.py]: That would require adding implementation for `GET` request for the same endpoint (I'll do that)

Comment (abhaasgoyal) [reply meorg_client/tests/test_client.py]: I implemented the `GET` endpoint, but seems to hang the response for existing model output id's. I'll get in touch with Gavin about this.

I added the necessary TODOs and skipped the test for getting details for model output IDs.

File: meorg_client/cli.py (hunk 5 above)
Comment (bschroeter) [reply meorg_client/cli.py:247]: OK, so the spec suggests that a boolean is returned, but which value references to which circumstance?

Is it True = created? or True = existing?

File: docs/cli.md [hunk 10]
```diff
@@ -93,6 +93,28 @@ modelevaluation.org/modelOutput/display/**kafS53HgWu2CDXxgC**
 
 This command will return an `$ANALYSIS_ID` upon success which is used in `analysis status`.
 
+### model output create
+
+To start an analysis for a given model output using the files provided, execute the following command:
```
Comment (bschroeter) [inline docs/cli.md:98]:
```suggestion
To create a model output, execute the following command:
```

File: docs/cli.md [hunk 11]
```diff
@@ -93,6 +93,28 @@ modelevaluation.org/modelOutput/display/**kafS53HgWu2CDXxgC**
...
+
+```shell
+meorg output create $MODEL_PROFILE_ID $EXPERIMENT_ID $MODEL_OUTPUT_NAME
+```
+
+Where `$MODEL_PROFILE_ID` and `$EXPERIMENT_ID` are found on the model profile details and corresponding experiment details pages in question. `$MODEL_OUTPUT_NAME` is the name for the newly created model output.
```
Comment (bschroeter) [inline docs/cli.md:104]:
```suggestion
Where `$MODEL_PROFILE_ID` and `$EXPERIMENT_ID` are found on the model profile and corresponding experiment details pages on modelevaluation.org. `$MODEL_OUTPUT_NAME` is a unique name for the newly created model output.
```

File: docs/cli.md [hunk 12]
```diff
@@ -93,6 +93,28 @@ modelevaluation.org/modelOutput/display/**kafS53HgWu2CDXxgC**
...
+
+### model output query
+
+Retrieve Model output details via `$MODEL_OUTPUT_ID`
+
+```shell
+meorg output query $MODEL_OUTPUT_ID
+```
+
+This command will print the `id` and `name` of the modeloutput, and JSON representation for the remaining metadata.
```
Comment (bschroeter) [inline docs/cli.md:116]: Does it dump the JSON? Do we want it to do that?

File: meorg_client/cli.py [hunk 13]
```diff
@@ -214,6 +214,71 @@ def analysis_start(id: str):
...
+    """
+    client = _get_client()
+
+    response = _call(client.model_output_query, model_id=model_id)
+
+    if client.success():
+        model_output_id = response.get("data").get("id")
+        name = response.get("data").get("name")
+        remaining_data = response.get("data")
+        remaining_data.pop("id")
```
Comment (bschroeter) [inline meorg_client/cli.py:274]: It took me a little while to figure out what was going on here. Why not leave the id and name in the JSON dict? I see no reason to pop them.

File: meorg_client/endpoints.py [hunk 14]
```diff
@@ -16,3 +16,7 @@
 # Analysis
 ANALYSIS_START = "modeloutput/{id}/start"
 ANALYSIS_STATUS = "analysis/{id}/status"
+
+# Model Outputs
+MODEL_OUTPUT_CREATE = "modeloutput"
+MODEL_OUTPUT_QUERY = "modeloutput"
```
Comment (bschroeter) [inline meorg_client/endpoints.py:22]:
```suggestion
MODEL_OUTPUT_QUERY = MODEL_OUTPUT_CREATE
```

File: meorg_client/tests/test_cli.py [hunk 15]
```diff
@@ -56,6 +56,32 @@ def test_list_endpoints(runner: CliRunner):
...
+        standalone_mode=False,
+    )
+
+    # TODO: Add test for model output query after API endpoint for GET resolved
+
+    assert result.exit_code == 0
+    assert type(result.return_value) is str  # The new model output
+
+
+@pytest.mark.skip(reason="Response timeout")
```
Comment (bschroeter) [inline meorg_client/tests/test_cli.py:75]: Why are you skipping this?

File: meorg_client/tests/test_client.py [hunk 16]
```diff
@@ -93,6 +93,28 @@ def test_list_endpoints(client: Client):
...
+    )
+    assert client.success()
+
+    model_output_id = response.get("data").get("modeloutput")
+    assert model_output_id is not None
+
+    # TODO: Add test for model output query after API endpoint for GET resolved
+
+
+@pytest.mark.skip(reason="Response timeout")
```
Comment (bschroeter) [inline meorg_client/tests/test_client.py:111]: Why are you skipping this?

**Review by bschroeter — CHANGES_REQUESTED** [review]

File: docs/cli.md (hunk 12 above)
Comment (abhaasgoyal) [reply docs/cli.md:116]: Hmm, I thought it could be useful to have all the model output details for the developer mode, so I have added a conditional check for that.

File: meorg_client/tests/test_cli.py (hunk 15 above)
Comment (abhaasgoyal) [reply meorg_client/tests/test_cli.py]: The get endpoint was not working previously, so I thought I should move it to another issue. However, on further analysis, I had to construct the URL correctly. I have refactored the necessary code for this

File: meorg_client/tests/test_client.py (hunk 3 above)
Comment (abhaasgoyal) [reply meorg_client/tests/test_client.py]: Update: The `GET` endpoint for Model output is working now, so I have added the necessary additional test for that.

File: meorg_client/constants.py [hunk 17]
```diff
@@ -17,3 +19,5 @@
 
 # Production URL
 MEORG_BASE_URL_PROD = "https://modelevaluation.org/api"
+
+IS_DEV_MODE = os.getenv("MEORG_DEV_MODE", "0") == "1"
```
Comment (bschroeter) [inline meorg_client/constants.py:23]: Strictly speaking, this is not a constant, as it can be changed external to the code. I will allow it for now, but I think it better if we had a utility function to detect dev-mode-ness.

*In utilities.py*
```python
def is_dev_mode():
    return os.getenv("MEORG_DEV_MODE", "0") == "1"
```

File: meorg_client/cli.py [hunk 18]
```diff
@@ -20,17 +22,16 @@ def _get_client() -> Client:
...
     """
     # Get the dev-mode flag from the environment, better than passing the dev flag everywhere.
-    dev_mode = os.getenv("MEORG_DEV_MODE", "0") == "1"
 
     credentials = mcu.get_user_data_filepath("credentials.json")
     credentials_dev = mcu.get_user_data_filepath("credentials-dev.json")
 
     # In dev mode and the configuration file exists
-    if dev_mode and credentials_dev.is_file():
+    if mcc.IS_DEV_MODE and credentials_dev.is_file():
```
Comment (bschroeter) [inline meorg_client/cli.py:30]: See below for comment on constants.

File: meorg_client/tests/test_cli.py [hunk 19]
```diff
@@ -56,6 +56,31 @@ def test_list_endpoints(runner: CliRunner):
...
+    result = runner.invoke(
+        cli.create_new_model_output,
+        [model_profile_id, experiment_id, model_output_name],
+        standalone_mode=False,
+    )
+
+    assert result.exit_code == 0
+    assert type(result.return_value) is str  # The new model output
+
+    test_model_output_query(runner, result.return_value)
```
Comment (bschroeter) [inline meorg_client/tests/test_cli.py:72]: This should use the `store` mechanism, rather than calling the next test. As it stands, `test_model_output_query` is called twice in the test suite.

Result is the same, the code is tested, but there is a cleaner way to do this.

File: meorg_client/client.py [hunk 20]
```diff
@@ -129,22 +130,29 @@ def _make_request(
...
         -------
         str
             URL.
         """
-        return urljoin(self.base_url + "/", endpoint).format(**kwargs)
+        # Add endpoint to base URL, interpolating url_path_fields
+        url_path = urljoin(self.base_url + "/", endpoint).format(**url_path_fields)
+        # Add URL parameters (if any)
+        if url_params:
+            url_path = f"{url_path}?{urlencode(url_params)}"
```
Comment (bschroeter) [inline meorg_client/client.py:154]: Nice

**Review by bschroeter — APPROVED** [review]
A few small suggestions, but in the interest of closing this PR we can shunt them to another issue.

Nice work.

File: meorg_client/tests/test_cli.py (hunk 19 above)
Comment (abhaasgoyal) [reply meorg_client/tests/test_cli.py:72]: `store` stores the environment variable `MEORG_MODEL_OUTPUT_ID`, which is separate from the result `model_output_id` generated above. In the interest of keeping the tests non-dependent as possible (so that `store` doesn't get overriden), I had separated the calls.
